esper/ext/mediator.py

```diff
@@ -0,0 +1,334 @@
+import selectors
+import socket
+import types
+from typing import Tuple
+import random
+import sys
+import traceback
+from logging import Logger
+from datetime import datetime
+
+from esper.ext.forwarder import TCPForwarder
+
+
+class MediatorShutdown(BaseException):
+    pass
+
+
+class Mediator(object):
+    selector = None
+    listener = None
+    log = None
+    _host = None
+    _port = None
+
+    _insecure_connection = None
+    _secure_connection = None
+    _secure_addr = None
+
+    _outbound_data = b''
+    _inbound_data = b''
+
+    @property
+    def host(self) -> str:
+        return self._host
+
+    @property
+    def port(self) -> int:
+        return self._port
+
+    def __init__(self, secure_conn=None, secure_addr=None, log=None):
+        self.selector = selectors.DefaultSelector()
+        self.log = log
+        self._host = '127.0.0.1'
+        self._port = self._get_random_unused_port()
+
+        self._secure_connection = secure_conn
+        self._secure_addr = secure_addr
+
+    def _get_random_unused_port(self, min_port=47000, max_port=57000) -> int:
+        '''Iterate over a range of ports and pick the first free port and return it '''
+
+        for count in range((max_port - min_port)):
+            port = random.randrange(min_port, max_port)
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as probe:
+                resp = probe.connect_ex((socket.gethostname(), port))
```

Comment:
Is this checking if port is currently in use ?

Reply:
yes. 
`socket.connect_ex()` will return an error code instead of exception. If there is a non-zero error code, it means it failed to connect, meaning there is no service running on that port.